src/providers/wfs/qgswfsdataitems.cpp

```diff
@@ -50,6 +55,76 @@ QgsWfsLayerItem::~QgsWfsLayerItem()
 {
 }
 
+QList<QMenu *> QgsWfsLayerItem::menus()
+{
+  QList<QMenu *> menus;
+
+  if ( mPath.startsWith( QLatin1String( "geonode:/" ) ) )
+  {
+    QAction *actionCopyStyle = new QAction( tr( "Copy Style" ), this );
+    connect( actionCopyStyle, &QAction::triggered, this, &QgsWfsLayerItem::copyStyle );
+
+    QMenu *menuStyleManager = new QMenu();
+    menuStyleManager->setTitle( tr( "Styles" ) );
+    menuStyleManager->addAction( actionCopyStyle );
+    menus << menuStyleManager;
+  }
+
+  return menus;
+}
+
+void QgsWfsLayerItem::copyStyle()
+{
+  QString layerUri = this->uri();
+
+  QgsGeoNodeConnection *connection = nullptr;
+  Q_FOREACH ( const QString &connName, QgsGeoNodeConnectionUtils::connectionList() )
+  {
+    connection = new QgsGeoNodeConnection( connName );
+    if ( layerUri.contains( connection->uri().uri() ) )
+      break;
+  }
+
+  if ( !connection )
+  {
+    QString errorMsg( QStringLiteral( "Cannot get style for layer %1" ).arg( this->name() ) );
+    QgsDebugMsg( " Cannot get style: " + errorMsg );
+#if 0
+    // TODO: how to emit message from provider (which does not know about QgisApp)
```

Comment:
@nyalldawson we need to find a decent solution for this use case, everybody needs it. I cannot think of anything better than a slot in qgis app that accepts a message and a level and connect it to something like `&QgisApp::authMessageOut` ( `void QgisApp::authMessageOut( const QString &message, const QString &authtag, QgsAuthManager::MessageLevel level )` ).